include/alpaka/block/shared/st/BlockSharedMemStMember.hpp

```diff
@@ -0,0 +1,146 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/block/shared/st/Traits.hpp>
+#include <alpaka/core/Assert.hpp>
+
+#include <type_traits>
+#include <cstdint>
+
+namespace alpaka
+{
+    namespace block
+    {
+        namespace shared
+        {
+            namespace st
+            {
+                namespace detail
+                {
+                    //#############################################################################
+                    //! Implementation of static block shared memory provider.
+                    template<unsigned int TDataAlignBytes = 4>
+                    class BlockSharedMemStMemberImpl
+                    {
+                        mutable unsigned int m_allocdBytes = 0;
+                        mutable uint8_t* m_mem;
+#ifndef NDEBUG
+                        const unsigned int m_capacity;
+#endif
+
+                        template<typename T>
+                        static constexpr size_t alignPitch()
+                        {
+                            return (sizeof(T)/TDataAlignBytes + (sizeof(T)%TDataAlignBytes>0))*TDataAlignBytes;
+                        }
+
+                    public:
+                        //-----------------------------------------------------------------------------
+#ifndef NDEBUG
+                        BlockSharedMemStMemberImpl(uint8_t* mem, unsigned int capacity) : m_mem(mem), m_capacity(capacity) {}
+#else
+                        BlockSharedMemStMemberImpl(uint8_t* mem, unsigned int) : m_mem(mem) {}
+#endif
+                        //-----------------------------------------------------------------------------
+                        BlockSharedMemStMemberImpl(BlockSharedMemStMemberImpl const &) = delete;
+                        //-----------------------------------------------------------------------------
+                        BlockSharedMemStMemberImpl(BlockSharedMemStMemberImpl &&) = delete;
+                        //-----------------------------------------------------------------------------
+                        auto operator=(BlockSharedMemStMemberImpl const &) -> BlockSharedMemStMemberImpl & = delete;
+                        //-----------------------------------------------------------------------------
+                        auto operator=(BlockSharedMemStMemberImpl &&) -> BlockSharedMemStMemberImpl & = delete;
+                        //-----------------------------------------------------------------------------
+                        /*virtual*/ ~BlockSharedMemStMemberImpl() = default;
+
+                        template <typename T>
+                        void alloc() const
+                        {
+                            uint8_t* buf = &m_mem[m_allocdBytes];
+                            new (buf) T();
+                            m_allocdBytes += alignPitch<T>();
+#if (defined ALPAKA_DEBUG_OFFLOAD_ASSUME_HOST) && (! defined NDEBUG)
```

Comment:
Is there some way to find out from the compiler if we are compiling for offload? Some macros? builtins?

Reply:
There are not portable OpenACC or OpenMP 5 things. The only way might be to check for things like `__CUDA_ARCH__`.

Reply:
Do we have a different portable way of signaling this error instead of an assertion? Exceptions, exit?

Reply:
`exit()` should be as forbidden on GPU as `assert()`,, also `exit()` would also require some message to be printed, which one may not want to routinely compile into the GPU Code event in debug mode. I am not sure about `throw`, I have the feeling, that this does not work either.

Reply:
`throw` is not allowed in CUDA kernels. Surprisingly, OpenMP 5 allows `throw` in target and parallel regions as long as the exception is caught inside the same region by the same thread. In the best case, not catching it would yield a runtime error and thus terminate (which calls exit and can lead to another issue, see https://github.com/ComputationalRadiationPhysics/picongpu/issues/3241 ), worst case a missing `try{}catch{}` causes a compile time error.

In any case, I would say `throw` is, apart from being not generally save, carrying the wrong semantics: You put `throw` where you might want to catch an error, not to report a bug in user code (here, exceeding the set shared mem limit).